core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocatorService.java

```diff
@@ -0,0 +1,274 @@
+
+public class FixedCountNodeAllocatorService
+        implements NodeAllocatorService
+{
+    private static final Logger log = Logger.get(FixedCountNodeAllocatorService.class);
+    private static final int MAXIMUM_ALLOCATIONS_PER_NODE = 1; // TODO make configurable?
+
+    private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("fixed-count-node-allocator"));
+    private final NodeScheduler nodeScheduler;
+
+    private final Set<FixedCountNodeAllocator> allocators = newConcurrentHashSet();
+    private final AtomicBoolean started = new AtomicBoolean();
+
+    @Inject
+    public FixedCountNodeAllocatorService(NodeScheduler nodeScheduler)
+    {
+        this.nodeScheduler = requireNonNull(nodeScheduler, "nodeScheduler is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.scheduleWithFixedDelay(() -> {
+                try {
+                    updateNodes();
+                }
+                catch (Throwable e) {
+                    // ignore to avoid getting unscheduled
+                    log.warn(e, "Error updating nodes");
+                }
+            }, 5, 5, TimeUnit.SECONDS);
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        executor.shutdownNow();
+    }
+
+    @VisibleForTesting
+    void updateNodes()
+    {
+        allocators.forEach(FixedCountNodeAllocator::updateNodes);
+    }
+
+    @Override
+    public NodeAllocator getNodeAllocator(Session session)
+    {
+        requireNonNull(session, "session is null");
+        return getNodeAllocator(session, MAXIMUM_ALLOCATIONS_PER_NODE);
+    }
+
+    @VisibleForTesting
+    NodeAllocator getNodeAllocator(Session session, int maximumAllocationsPerNode)
+    {
+        FixedCountNodeAllocator allocator = new FixedCountNodeAllocator(session, maximumAllocationsPerNode);
+        allocators.add(allocator);
+        return allocator;
+    }
+
+    private class FixedCountNodeAllocator
+            implements NodeAllocator
+    {
+        private final Session session;
+        private final int maximumAllocationsPerNode;
+
+        @GuardedBy("this")
+        private final Map<Optional<CatalogName>, NodeSelector> nodeSelectorCache = new HashMap<>();
+
+        @GuardedBy("this")
+        private final Map<InternalNode, Integer> allocationCountMap = new HashMap<>();
+
+        @GuardedBy("this")
+        private final LinkedList<PendingAcquire> pendingAcquires = new LinkedList<>();
+
+        public FixedCountNodeAllocator(
+                Session session,
+                int maximumAllocationsPerNode)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.maximumAllocationsPerNode = maximumAllocationsPerNode;
+        }
+
+        @Override
+        public synchronized ListenableFuture<InternalNode> acquire(NodeRequirements requirements)
```

Comment:
ok. So it looks like you try to acquire node for a task with given requirements.

This has limitations as it's not possible to move retriable task after it has been scheduled. If you look at `ExecutionPolicy` and `SourcePartitionedScheduler` you will see that they are aware of objects they schedule (e.g. stages or splits). `ExecutionPolicy` keeps references to running stages.

If you implemented `NodeAllocator` is a similar fashion then you would have handles to `Tasks` and could move them around when there comes an opportunity.

Do you think it would make sense to be able to move existing task (be retrying it)?

Reply:
Every tasks attempt (on retry) is scheduled independently and can land on a different node.
Currently when we make after we made decision to schedule task on a given node we no longer can move it, even if it not started execution yet. Maybe it would make sense to allow for that but it feels like a massive change - not to be done before MVP. WDYT @arhimondr

Reply:
For the future. This might potentially not scale as `tasks` are essentially kind of splits with approach. For each `task` we create a future and inform about assignment via this future and some executor. Perhaps this should use more batch-like approach similar to `SourcePartitionedScheduler`

Reply:
@sopel39 Could you please elaborate what do you mean by "moving tasks"? Currently it is possible to "move" a task by killing an existing task on "Node A" and scheduling another copy of the same task on "Node B".